File: app/src/main/java/cat/mvmike/minimalcalendarwidget/domain/entry/Instance.java
// Copyright (c) 2016, Miquel Martí <miquelmarti111@gmail.com>
// See LICENSE for licensing information

package cat.mvmike.minimalcalendarwidget.domain.entry;

import android.provider.CalendarContract;

import java.time.Instant;


public final class Instance {

    public static final String[] FIELDS = {
        CalendarContract.Instances.BEGIN,
        CalendarContract.Instances.END,
        CalendarContract.Instances.START_DAY,
        CalendarContract.Instances.END_DAY
    };

    private static final int MILLIS_IN_A_DAY = 86_400_000;

    private final Instant start;

    private final Instant end;

    private final boolean allDay;

    public Instance(final long epochMilliStart, final long epochMilliEnd, final int julianStartDay, final int julianEndDate) {
        this.start = Instant.ofEpochMilli(epochMilliStart);
        this.end = Instant.ofEpochMilli(epochMilliEnd);
        this.allDay = computeAllDay(start, end, julianStartDay, julianEndDate);
    }

    public Instant getStart() {
        return start;
    }

    public Instant getEnd() {
        return end;
    }

    public boolean isAllDay() {
        return allDay;
    }

    private static boolean computeAllDay(final Instant start, final Instant end, final int julianStartDay, final int julianEndDate) {
        return ((end.toEpochMilli() - start.toEpochMilli()) % MILLIS_IN_A_DAY == 0)
            && ((end.toEpochMilli() - start.toEpochMilli()) / MILLIS_IN_A_DAY == (julianEndDate - julianStartDay) + 1);
    }
}

File: app/src/main/kotlin/cat/mvmike/minimalcalendarwidget/domain/entry/Instance.kt
// Copyright (c) 2016, Miquel Martí <miquelmarti111@gmail.com>
// See LICENSE for licensing information
package cat.mvmike.minimalcalendarwidget.domain.entry

import android.content.Context
import cat.mvmike.minimalcalendarwidget.infrastructure.SystemResolver
import cat.mvmike.minimalcalendarwidget.infrastructure.config.ClockConfig
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId

data class Instance(
    val eventId: Int,
    val start: Instant,
    val end: Instant,
    val zoneId: ZoneId,
    val isDeclined: Boolean
) {
    // take out 5 milliseconds to avoid erratic behaviour events that end at midnight
    fun isInDay(day: LocalDate): Boolean {
        val instanceStartLocalDate = LocalDateTime.ofInstant(start, zoneId).toLocalDate()
        val instanceEndLocalDate = LocalDateTime.ofInstant(end.minusMillis(5), zoneId).toLocalDate()
        return !instanceStartLocalDate.isAfter(day)
                && !instanceEndLocalDate.isBefore(day)
    }
}

fun getInstances(context: Context, from: LocalDate, to: LocalDate): Set<Instance> {
    return when (SystemResolver.isReadCalendarPermitted(context)) {
        false -> HashSet()
        true -> {
            val systemZoneId = ClockConfig.getSystemZoneId()
            SystemResolver.getInstances(
                context = context,
                begin = from.toStartOfDayInEpochMilli(systemZoneId),
                end = to.toStartOfDayInEpochMilli(systemZoneId)
            )
        }
    }
}

private fun LocalDate.toStartOfDayInEpochMilli(zoneId: ZoneId) =
    this.atStartOfDay(zoneId).toInstant().toEpochMilli()
